omegaconf/basecontainer.py

```diff
@@ -531,8 +531,12 @@ def _set_item_impl(self, key: Any, value: Any) -> None:
         if isinstance(value, Node):
             do_deepcopy = not self._get_flag("no_deepcopy_set_nodes")
             if not do_deepcopy and isinstance(value, Box):
+                # Detect the special case where value has no root:
+                value_has_root = (
+                    isinstance(value, Container) or value._get_parent() is not None
+                )
                 # if value is from the same config, perform a deepcopy no matter what.
-                if self._get_root() is value._get_root():
+                if value_has_root and self._get_root() is value._get_root():
```

Comment:
Normally, every node should have a root. The "root" of a node is a top-level `Container` that owns the node. In other words, the root is the top of the config tree.

In example from issue #1087, the call here to `value._get_root()` causes an `AssertionError`. This happens because `value` is a temporary `UnionNode` value that has no parent (and thus has no root).

Reply:
Both approaches seems okay to me.
Also, please add a test for unsafe_merge similar in this case (similar to the repro in the bug report).

Reply:
Make sure that your proposed logic does not treat the nodes as being from the same config (see comment in 538).

Reply:
I've added the test.

> Make sure that your proposed logic does not treat the nodes as being from the same config 

Yup. If `value_has_root` is `False` then `value` is a temporary node and does not come from the same config as `self`.